### PR TITLE
EdkRepo: List Repos Should Handle Combos That Do Not Reference a Branch

### PR DESCRIPTION
**Pull request by ashedesimone**
The list repos command will fail to process if any of the available manifest files contain a combo which does not reference a branch. Update the command to by-pass this case and to identify if a commit or a tag is used in a combo. The output has been updated to show all branches, commits or tags in a given repo that are used and which combo they are used in.

Fixes: #167

### CODE REVIEW

File: edkrepo/commands/list_repos_command.py [hunk 1]
```diff
@@ -202,38 +211,102 @@ def run_command(self, args, config):
                 branches.insert(0, 'main')
 
             #For each interesting branch in the current git repository...
-            for branch in branches:
-                branch_data = { 'name': branch, 'projects': [] }
+            #for branch in branches:
+                #branch_data = { 'name': branch, 'projects': [] }
```
Comment (ashedesimone) [inline edkrepo/commands/list_repos_command.py:215]: Remove this commented out code

File: edkrepo/commands/list_repos_command.py [hunk 2]
```diff
@@ -202,38 +211,102 @@ def run_command(self, args, config):
...
+                if ref_type[0]:
+                    branch_data = { 'name': ref_type[0], 'projects': [] }
+                else: 
+                    branch_data = None
+                if ref_type[1]:
+                    commit_data = { 'name': ref_type[1], 'projects': [] }
+                else:
+                    commit_data = None
+                if ref_type[2]:
+                    tag_data = { 'name':ref_type[2], 'projects': [] }
```
Comment (ashedesimone) [inline edkrepo/commands/list_repos_command.py:232]: add a space before ref_type

File: edkrepo/commands/list_repos_command.py [hunk 3]
```diff
@@ -202,38 +211,102 @@ def run_command(self, args, config):
...
                     combo_list = [c.name for c in manifests[project_name].combinations]
                     if args.archived:
                         combo_list.extend([c.name for c in manifests[project_name].archived_combinations])
                     for combo in combo_list:
                         sources = manifests[project_name].get_repo_sources(combo)
                         for source in sources:
-                            if self.get_repo_url(source.remote_url) == repo and source.branch == branch:
-                                combos.append(combo)
+                            if self.get_repo_url(source.remote_url) == repo: #and (source.branch == ref_type[0] or source.commit == ref_type[1] or source.tag == ref_type[2]):
+                                #combos.append(combo)
```
Comment (ashedesimone) [inline edkrepo/commands/list_repos_command.py:248]: remove commented out code

File: edkrepo/commands/list_repos_command.py [hunk 4]
```diff
@@ -202,38 +211,102 @@ def run_command(self, args, config):
...
+                                if ref_type[0]:
+                                    if ref_type[0] == source.branch:
+                                #if source.branch == ref_type[0]:
```
Comment (ashedesimone) [inline edkrepo/commands/list_repos_command.py:251]: remove commented out code

File: edkrepo/commands/list_repos_command.py [hunk 5]
```diff
@@ -202,38 +211,102 @@ def run_command(self, args, config):
...
+                                      branch_combos.append(combo)
+                                      add_branch = True
+                                #if source.commit == ref_type[1]:
```
Comment (ashedesimone) [inline edkrepo/commands/list_repos_command.py:254]: remove commented out code

**Comment by ndhaller** [comment]
Functional testing 3.1.1 w/ this patch, all of my test manifest-repos were able to be processed, though patchSet are excluded from the scope of this commit.

test system: Python 3.9.13, git version 2.39.1.windows.1, Windows 11

File: edkrepo/commands/list_repos_command.py [hunk 6]
```diff
@@ -183,7 +184,12 @@ def run_command(self, args, config):
                     sources = manifests[project_name].get_repo_sources(combo)
                     for source in sources:
                         if self.get_repo_url(source.remote_url) == repo:
-                            branches.add(source.branch)
+                            if source.branch:
```
Comment (ndhaller) [inline edkrepo/commands/list_repos_command.py:187]: I made a test Combination that uses commit and tag and branch at the same time.
When the combo is checked out, the priority is supposed to be commit > tag > branch

```
    <Combination name="sandbox-listRepos-test2-commit-tag-branch" description="a combo for testing the list-repos command with commit > tag > branch priority">
      <Source localRoot="edkrepo" remote="EdkrepoRepo" commit="31bb065c7f288437e2823a95606928325d6b2467" tag="edkrepo-v3.1.1" branch="main" />
      <Source localRoot="edkrepo-manifest" remote="EdkrepoManifestRepo" commit="498fc403d3ab9fd96fbd9ac3ca78dd01e5ead466" branch="main" />
    </Combination>
```

The edkrepo list-repos command only lists this Combination by the branch name.

```
edkrepo - [https://github.com/tianocore/edk2-edkrepo]
Branches:
  main
    EdkRepo                                      : *(main)*
    Edkrepo-Sandbox2                             : *(sandbox-main2)*
                                                    (sandbox-listRepos-test2-commit-tag-branch)
edkrepo-manifest - [https://github.com/tianocore/edk2-edkrepo-manifest]
Branches:
  main
    EdkRepoManifest                              : *(main)*
    Edkrepo-Sandbox2                             : *(sandbox-main2)*
                                                    (sandbox-listRepos-test2-commit-tag-branch)
```

Should "edkrepo list-repos" follow the same priority of commit > tag > branch?

**Review by ndhaller — COMMENTED** [review]
line 471
```
old_repo_url = self.repo_names[name][0]
```

should be:
```
old_repo_url = self.repo_names[best_name][0]
```

when running 'edkrepo list-repos' multiple times, occasionally this failure occurs:

'repoName1' is assigned to a repository with four instances.
defaultdict(<class 'int'>, {'repoName1': 4})

'repoName1' wants to be changed to a repository that uses it with thirty-two instances.
defaultdict(<class 'int'>, {'repoName1': 32, 'repoName2': 2})

'repoName2' is not 'best_name' because it only is used in a minority of cases, but it is the value stored in 'name' since it is last accessed.

```
Traceback (most recent call last):
  File "C:\Python39\lib\site-packages\edkrepo\edkrepo_cli.py", line 183, in main
    command.run_command(command_name, parsed_args, config)
  File "C:\Python39\lib\site-packages\edkrepo\commands\composite_command.py", line 38, in run_command
    return command.run_command(args, config)
  File "C:\Python39\lib\site-packages\edkrepo\commands\list_repos_command.py", line 161, in run_command
    self.generate_repo_names(repo_urls, manifests, args.archived)
  File "C:\Python39\lib\site-packages\edkrepo\commands\list_repos_command.py", line 399, in generate_repo_names
    self.__repo_name_worker(repo_url, manifests, archived)
  File "C:\Python39\lib\site-packages\edkrepo\commands\list_repos_command.py", line 471, in __repo_name_worker
    old_repo_url = self.repo_names[name][0]
KeyError: 'repoName2'
```

File: edkrepo/commands/list_repos_command.py [hunk 7]
```diff
@@ -202,38 +211,97 @@ def run_command(self, args, config):
...
+                                    if ref_type[1] == source.commit:
+                                      commit_combos.append(combo)
+                                      add_commit = True
+                                if ref_type[2]:
+                                    if ref_type[2] == source.tag:
+                                      tag_combos.append(combo)
+                                      add_tag = True
                                 break
-                    if len(combos) > 0:
+                    if len(branch_combos or commit_combos or tag_combos) > 0:
```
Comment (nate-desimone) [inline edkrepo/commands/list_repos_command.py:259]: Needs to be

`if len(a) > 0 or len(b) > 0 or len(c) > 0:`

File: edkrepo/commands/list_repos_command.py (hunk 6 above)
Comment (nate-desimone) [reply edkrepo/commands/list_repos_command.py:187]: We should probably show all the commits/tags/branches that are being used, but only put the "*" and green color around the one that is used by default.

**Review by nate-desimone — CHANGES_REQUESTED** [review]

**Comment by ashedesimone** [comment]
> Functional testing 3.1.1 w/ this patch, all of my test manifest-repos were able to be processed, though patchSet are excluded from the scope of this commit.
> 
> test system: Python 3.9.13, git version 2.39.1.windows.1, Windows 11

The omission of PatchSets is by design; there is no requirement that patchset names be unique amongst the set of all possible manifest files which introduces the possibility that two usasges of the same patchset name are not actually using the same patchset.

**Review by nate-desimone — APPROVED** [review]

**Pull request merged**